src/test/scala/com/dwolla/sqlserver/init/ArbitraryRefinedTypes.scala

```diff
@@ -0,0 +1,61 @@
+package com.dwolla.sqlserver.init
+
+import cats._
+import cats.syntax.all._
+import com.dwolla.sqlserver.init.repositories.ReservedWords
+import eu.timepit.refined.api.{Refined, Validate}
+import eu.timepit.refined.refineV
+import org.scalacheck.{Arbitrary, Gen, Shrink}
+
+trait ArbitraryRefinedTypes {
+  def refinedConst[A] = new PartiallyAppliedRefinedConst[A]
+
+  implicit val shrinkSqlIdentifier: Shrink[SqlIdentifier] = Shrink.shrinkAny
+  def genSqlIdentifier[F[_] : Applicative]: Gen[F[SqlIdentifier]] =
+    for {
+      initial <- Gen.alphaChar
+      len <- Gen.chooseNum(0, 27) // see comment on SqlIdentifier
+      tail <- Gen.stringOfN(len, Gen.oneOf(Gen.alphaChar, Gen.numChar, Gen.const('_')))
+      identifier = s"$initial$tail"
+      if !ReservedWords.contains(identifier)
+      refined <- refineV[SqlIdentifierPredicate](s"$initial$tail").fold(_ => Gen.fail, Gen.const)
+    } yield refined.pure[F]
+  implicit val arbSqlIdentifier: Arbitrary[SqlIdentifier] = Arbitrary(genSqlIdentifier[Id])
+
+  implicit val shrinkSqlServerUser: Shrink[SqlServerUser] = Shrink.shrinkAny
+  def genSqlServerUser[F[_] : Applicative]: Gen[F[SqlServerUser]] =
+    for {
+      initial <- Gen.alphaChar
+      len <- Gen.chooseNum(0, 63)
+      tail <- Gen.stringOfN(len, Gen.oneOf(Gen.alphaChar, Gen.numChar, Gen.const('_')))
+      user = s"$initial$tail"
+      if !ReservedWords.contains(user)
+      refined <- refineV[SqlServerUserPredicate](s"$initial$tail").fold(_ => Gen.fail, Gen.const)
+    } yield refined.pure[F]
+  implicit val arbSqlServerUser: Arbitrary[SqlServerUser] = Arbitrary(genSqlServerUser[Id])
+
+  implicit val shrinkGeneratedPassword: Shrink[GeneratedPassword] = Shrink.shrinkAny
+  def genGeneratedPassword[F[_] : Applicative]: Gen[F[GeneratedPassword]] = {
+    val allowedPunctuation: List[Char] = """! " # $ % & ( ) * + , - . / : < = > ? @ [ ] ^ _ { | } ~ """.replaceAll(" ", "").toList
+    val allowedPunctuationGen: Gen[Char] = Gen.oneOf(allowedPunctuation)
+    val allowedCharacters: Gen[Char] = Gen.oneOf(Gen.alphaChar, Gen.numChar, allowedPunctuationGen)
+
+    for {
+      length <- Gen.chooseNum(8, 128)
+      password <- Gen.stringOfN(length, allowedCharacters)
+        .suchThat(_.exists(_.isDigit))
```

Comment:
These `.suchThat` commands are ok, but I don't think they're doing anything for us, because the `allowedCharacters` generator already enforces these restrictions. (Which is good, because we want the generator to select characters from a limited set of characters, as opposed to selecting from a wide set of characters and filtering the resulting stream, since post-generation filtering is limited in the number of items it will reject before failing the entire thing.)

Reply:
I wish I had saved the seed for the failing test I had. I don't think the generator guarantees that the string includes at least 1 letter, number, & symbol.

Reply:
Oh, that's a good point—we can make that happen, though! Something like this should work:

```scala
def stringContainingAtLeastOneOfEachClass(length: Int, charClassGenerators: Gen[Char]*): Gen[String] = {
  assert(length >= charClassGenerators.length, "Output string length must be greater than or equal to the number of character classes")

  for {
    oneFromEachClass <- Gen.sequence(charClassGenerators).map(_.asScala)
    filler <- Gen.listOfN(length - charClassGenerators.length, Gen.choose(0, charClassGenerators.length - 1).flatMap(i => charClassGenerators(i)))
    chars <- Gen.long.flatMap(new Random(_).shuffle(oneFromEachClass ++ filler))
  } yield chars.mkString
}
```

and then

```scala
password <- stringContainingAtLeastOneOfEachClass(length, Gen.alphaChar, Gen.numChar, allowedPunctuationGen)
```